pennylane/transforms/split_non_commuting.py

```diff
@@ -167,75 +227,451 @@ def circuit0(x):
-            new_tape = tape.__class__(
-                tape.operations, (tape.measurements[i] for i in indices), shots=tape.shots
+    # Special case for a single measurement of a Sum or Hamiltonian, in which case
+    # the grouping information can be computed and cached in the observable.
+    if (
+        len(tape.measurements) == 1
+        and isinstance(tape.measurements[0], ExpectationMP)
+        and isinstance(tape.measurements[0].obs, (Hamiltonian, Sum))
+        and (grouping_strategy is not None or tape.measurements[0].obs.grouping_indices is not None)
+    ):
+        return _split_ham_with_grouping(tape)
+
+    single_term_obs_mps, offsets = _split_all_multi_term_obs_mps(tape)
+
+    if grouping_strategy is None:
+        measurements = list(single_term_obs_mps.keys())
+        tapes = [tape.__class__(tape.operations, [m], shots=tape.shots) for m in measurements]
+        return tapes, partial(
+            _processing_fn_no_grouping,
+            single_term_obs_mps=single_term_obs_mps,
+            offsets=offsets,
+            shots=tape.shots,
+        )
+
+    if (
+        grouping_strategy == "naive"
+        or grouping_strategy == "default"
+        and any(
+            isinstance(m, ExpectationMP) and isinstance(m.obs, (LinearCombination, Hamiltonian))
+            for m in tape.measurements
+        )
+    ):
+        # This is a loose check to see whether naive grouping or pauli grouping should be used,
+        # which does not necessarily make perfect sense but consistent with the old decision
+        # logic in `LegacyDevice.batch_transform`. The premise is that pauli grouping is
+        # classically expensive but produces fewer tapes, whereas naive grouping is classically
+        # faster to compute, but inefficient quantum-wise. If this transform is to be added to a
+        # device's `preprocess`, it will be performed for every circuit execution, which can get
+        # very expensive if there is a large number of observables. The reasoning here is, large
+        # Hamiltonians typically come in the form of a `LinearCombination` or `Hamiltonian`, so
+        # if we see one of those, use naive grouping to be safe. Otherwise, use pauli grouping.
+        return _split_using_naive_grouping(tape, single_term_obs_mps, offsets)
+
+    return _split_using_pauli_grouping(tape, single_term_obs_mps, offsets)
+
+
+def _split_ham_with_grouping(tape: qml.tape.QuantumScript):
+    """Splits a tape measuring a single Hamiltonian or Sum and group commuting observables."""
+
+    obs = tape.measurements[0].obs
+    if obs.grouping_indices is None:
+        obs.compute_grouping()
+
+    coeffs, obs_list = obs.terms()
+
+    # The constant offset of the Hamiltonian, typically arising from Identity terms.
+    offset = 0
+
+    # A dictionary for measurements of each unique single-term observable, mapped to the
+    # indices of the original measurements it belongs to, and its coefficient.
```

Comment:
```suggestion
    # indices of the original measurements it belongs to, and its coefficient.
```
```suggestion
    # indices of the original measurements it belongs to, its coefficient,
    # the group idx, and the size of that group
```

I must say I'm a little confused by `group_size`. It seems to be more like the index of that measurement into the group, and not the overall size of that group.  Because the `group_size` changes after it's placed into `single_term_obs_mps`.  So obviously the fourth entry to `single_term_obs_mps[new_mp]` isn't the size of the group. It's the number of things that were already in the group when it got processed.

This might be a simple naming confusion thing.

Reply:
The `group_size` is the number of observables to be measured in each group, this is important to interpret the result dimensions. If `group_size==1`, the result returned from that tape would have the dimension corresponding to the number of measurements squeezed out.